### PR TITLE
fix: mergeGateways panics when restarting control plane

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/2282

### CODE REVIEW

**Comment by zirain** [comment]
can you add a (e2e)test for this to avoid regression in the future?

**Comment by Xunzhuo** [comment]
Can this be a follow up ? @zirain

**Comment by Xunzhuo** [comment]
Okay, let me update it later

**Comment by Xunzhuo** [comment]
I still think we can start working on e2e after https://github.com/envoyproxy/gateway/issues/2029#issuecomment-1808623065

let us fix the bug first @zirain

**Review by zirain — APPROVED** [review]

File: internal/provider/kubernetes/predicates.go [hunk 1]
```diff
@@ -179,7 +179,10 @@ func (r *gatewayAPIReconciler) validateServiceForReconcile(obj client.Object) bo
 	// Only merged gateways will have this label, update status of all Gateways under found GatewayClass.
 	gclass, ok := labels[gatewayapi.OwningGatewayClassLabel]
 	if ok {
-		res, _ := r.resources.GatewayAPIResources.Load(gclass)
+		res, ok := r.resources.GatewayAPIResources.Load(gclass)
+		if !ok {
+			return false
```
Comment (zzjin) [inline internal/provider/kubernetes/predicates.go:184]: If return false here, the service here(and deploy below) will always return false, witch means the `MergeGateways` do not work anymore.

**Comment by Xunzhuo** [comment]
Let me double check it before merging it.

File: internal/provider/kubernetes/predicates.go [hunk 2]
```diff
@@ -179,7 +179,10 @@ func (r *gatewayAPIReconciler) validateServiceForReconcile(obj client.Object) bo
...
+		}
 		for _, gw := range res.Gateways {
 			gw := gw
 			r.statusUpdateForGateway(ctx, gw)
```
Comment (cnvergence) [inline internal/provider/kubernetes/predicates.go:188]:
```suggestion
		}
	if res != nil && len(res.Gateways) > 0 {
		for _, gw := range res.Gateways {
			gw := gw
			r.statusUpdateForGateway(ctx, gw)
			}
		}
```

File: internal/provider/kubernetes/predicates.go [hunk 3]
```diff
@@ -312,7 +315,10 @@ func (r *gatewayAPIReconciler) validateDeploymentForReconcile(obj client.Object)
...
 	gclass, ok := labels[gatewayapi.OwningGatewayClassLabel]
 	if ok {
-		res, _ := r.resources.GatewayAPIResources.Load(gclass)
+		res, ok := r.resources.GatewayAPIResources.Load(gclass)
+		if !ok {
+			return false
+		}
 		for _, gtw := range res.Gateways {
 			gtw := gtw
 			r.statusUpdateForGateway(ctx, gtw)
```
Comment (cnvergence) [inline internal/provider/kubernetes/predicates.go:324]:
```suggestion
		}
	if res != nil && len(res.Gateways) > 0 {
		for _, gw := range res.Gateways {
			gw := gw
			r.statusUpdateForGateway(ctx, gw)
			}
		}
```

File: internal/provider/kubernetes/predicates.go (hunk 2 above)
Comment (cnvergence) [reply internal/provider/kubernetes/predicates.go:188]: we will need to add this check

**Review by cnvergence — CHANGES_REQUESTED** [review]
@Xunzhuo ptal

File: internal/provider/kubernetes/predicates.go [hunk 4]
```diff
@@ -179,10 +179,16 @@ func (r *gatewayAPIReconciler) validateServiceForReconcile(obj client.Object) bo
...
+		res, ok := r.resources.GatewayAPIResources.Load(gclass)
+		if !ok {
+			return false
+		}
+	if res != nil && len(res.Gateways) > 0 {
 		for _, gw := range res.Gateways {
 			gw := gw
 			r.statusUpdateForGateway(ctx, gw)
+			}
+		}
```
Comment (cnvergence) [inline internal/provider/kubernetes/predicates.go:191]: suggestion missed that
```suggestion

```

File: internal/provider/kubernetes/predicates.go [hunk 5]
```diff
@@ -312,10 +318,16 @@ func (r *gatewayAPIReconciler) validateDeploymentForReconcile(obj client.Object)
...
+		if !ok {
+			return false
+		}
+	if res != nil && len(res.Gateways) > 0 {
+		for _, gw := range res.Gateways {
+			gw := gw
+			r.statusUpdateForGateway(ctx, gw)
+			}
+		}
 		}
```
Comment (cnvergence) [inline internal/provider/kubernetes/predicates.go:331]:
```suggestion
```

Comment (cnvergence) [reply internal/provider/kubernetes/predicates.go:331]: same

**Review by cnvergence — APPROVED** [review]

**Review by cnvergence — APPROVED** [review]
lgtm now

**Review by zirain — APPROVED** [review]

**Pull request merged**